src/cdomain/value/cdomains/intDomain.ml

```diff
@@ -1337,137 +1366,108 @@ module BitfieldFunctor (Ints_t : IntOps.IntOps): SOverflow with type int_t = Int
-  let get_bit mask pos = ((Ints_t.to_int mask) lsr pos) land 1 = 1 in
-
-  (* Iterate from Most Significant Bit (MSB) to Least Significant Bit (LSB) *)
-  for i = bit_length - 1 downto 0 do
-    let bit1_zero = get_bit z1 i in
-    let bit1_one = get_bit o1 i in
-    let bit2_zero = get_bit z2 i in
-    let bit2_one = get_bit o2 i in
-
-    (* Check if bits at position i are both known *)
-    if (bit1_zero || bit1_one) && (bit2_zero || bit2_one) then
-      if bit1_zero && bit2_one then begin
-        result := if strict then Less else LessOrEqual;
-        raise Exit
-      end else if bit1_one && bit2_zero then begin
-        result := if strict then Greater else GreaterOrEqual;
-        raise Exit
-      end else if (bit1_one = bit2_one) && (bit1_zero = bit2_zero) then
-        () (* Equal bits, continue checking lower bits *)
-      else
-        result := Unknown (* Unknown bit situation, stop *)
-    else
-      result := Unknown; 
-      raise Exit
-  done;
+    else if not (leq x y || leq y x) then of_bool ik true
+    else BArith.top_bool
 
-  (* Handle sign bit adjustment if signed *)
-  if signed && !result <> Unknown then
-    match !result with
-    | Less when get_bit o1 sign_bit_position <> get_bit o2 sign_bit_position -> result := Greater
-    | Greater when get_bit o1 sign_bit_position <> get_bit o2 sign_bit_position -> result := Less
-    | _ -> ();
-  else ();
+  let le ik x y = failwith "Not implemented"
 
-  (* Handle non-strict inequalities for unknowns *)
-  if not strict && !result = Unknown then begin
-    if (Ints_t.logand z1 o2) =  Ints_t.zero then result := LessOrEqual
-    else if (Ints_t.logand o1 z2) = Ints_t.zero then result := GreaterOrEqual
-  end;
-  !result
+  let ge ik x y = failwith "Not implemented"
 
-  let ge ik x y = if compare_bitfields x y = GreaterOrEqual then of_bool ik true else BArith.topbool
+  let gt ik x y =failwith "Not implemented"
 
-  let le ik x y = if compare_bitfields x y = LessOrEqual then of_bool ik true else BArith.topbool
+  let lt ik x y =failwith "Not implemented"
 
-  let gt ik x y = if compare_bitfields x y = Greater then of_bool ik true else BArith.topbool
+  let invariant_ikind e ik (z,o) = 
+    let range = range ik (z,o) in
+    IntInvariant.of_interval e ik range
 
-  let lt ik x y = if compare_bitfields x y = Less then of_bool ik true else BArith.topbool
+  let starting ?(suppress_ovwarn=false) ik n = 
+    if Ints_t.compare n Ints_t.zero >= 0 then
+      (* sign bit can only be 0, as all numbers will be positive *)
+      let signBitMask = Ints_t.shift_left Ints_t.one (Size.bit ik - 1) in
+      let zs = BArith.one_mask in
+      let os = Ints_t.logand (Ints_t.lognot signBitMask) BArith.one_mask in
+      (norm ~suppress_ovwarn ik @@ (zs,os))
+    else 
+      (norm ~suppress_ovwarn ik @@ (top ()))
 
-  let invariant_ikind e ik = 
-    M.trace "bitfield" "invariant_ikind";
-    failwith "Not implemented"
+  let ending ?(suppress_ovwarn=false) ik n =
```

Comment:
What do you mean with `bf`? The arguments to this function are `ik` and `n`. 

I think `ending ik n` should return the bitmasks corresponding to all values `ik` could have, and are less than `n`.
So in mathematical form the result would be $$\text{JOIN } \\{ \text{ofInt } v. v \in  \text{values } \text{ik } \land v < n \\}$$

I don't think this can be computed efficiently, so I just did the case where $n<0$. Where its guaranteed that the sign bit is 1.

Reply:
Ok, I think this should work out, I had to write it down with pen and paper. For the case when $n > 0$ I was thinking about cases were we could output a bitfield more precise then top, but I guess this isn't possible since the join of the set would include a number that is possibly greater then the upper bound. Suppose we have an ik range of range(signed char) and n = 5 := 0101 then JOIN ... = 0TTT but 0111 >0101.